core/src/main/java/org/apache/calcite/plan/volcano/OptimizeTask.java

```diff
@@ -0,0 +1,237 @@
+   */
+  static class OptRelNodeTask extends OptimizeTask {
+    final RelNode node;
+    int nextId = 0; // next child index
+
+    OptRelNodeTask(RelNode node) {
+      super(node);
+      this.node = node;
+    }
+
+    @Override boolean hasSubTask() {
+      int size = node.getInputs().size();
+      while (nextId < size) {
+        RelSubset subset = (RelSubset) node.getInput(nextId);
+        if (subset.taskState == null) {
+          // not yet scheduled
+          return true;
+        } else {
+          // maybe a cycle if it is not completed
+          nextId++;
+        }
+      }
+
+      return false;
+    }
+
+    @Override OptimizeTask nextSubTask() {
+      RelNode child = node.getInput(nextId++);
+      return new OptSubsetTask((RelSubset) child);
+    }
+
+    @Override void execute() {
+      if (!(node instanceof PhysicalNode)
+          || ((PhysicalNode) node).getDeriveMode() == DeriveMode.PROHIBITED
+          || !planner.isSeedNode(node)) {
+        LOGGER.debug("Completed task(id={}) for {}", id, node);
+        return;
+      }
+
+      PhysicalNode rel = (PhysicalNode) node;
+      DeriveMode mode = rel.getDeriveMode();
+      int arity = node.getInputs().size();
+      // for OMAKASE
+      List<List<RelTraitSet>> inputTraits = new ArrayList<>(arity);
+
+      for (int i = 0; i < arity; i++) {
+        int childId = i;
+        if (mode == DeriveMode.RIGHT_FIRST) {
+          childId = arity - i - 1;
+        }
+
+        RelSubset input = (RelSubset) node.getInput(childId);
+        RelTraitSet required = input.getTraitSet();
+        List<RelTraitSet> traits = new ArrayList<>();
+        inputTraits.add(traits);
+
+        final int numSubset = input.set.subsets.size();
+        for (int j = 0; j < numSubset; j++) {
+          RelSubset subset = input.set.subsets.get(j);
+          if (!subset.isDelivered()
+              // TODO: should use matching type to determine
+              || required.equals(subset.getTraitSet())) {
+            continue;
+          }
+
+          if (mode == DeriveMode.OMAKASE) {
+            traits.add(subset.getTraitSet());
+          } else {
+            RelNode newRel = rel.derive(subset.getTraitSet(), childId);
+            if (newRel != null) {
+              RelSubset relSubset = planner.register(newRel, node);
+              assert relSubset.set == planner.getSubset(node).set;
+            }
+          }
+        }
+
+        if (mode == DeriveMode.LEFT_FIRST
+            || mode == DeriveMode.RIGHT_FIRST) {
+          break;
+        }
+      }
+
+      if (mode == DeriveMode.OMAKASE) {
+        List<RelNode> relList = rel.derive(inputTraits);
+        for (RelNode relNode : relList) {
+          planner.register(relNode, node);
+        }
+      }
+
+      LOGGER.debug("Completed task(id={}) for {}", id, node);
+    }
+
+    @Override public String toString() {
+      return "#" + id + ":OptRelNodeTask{ " + node + " }";
+    }
+  }
+
+  /**
+   * Task for optimizing RelSubset.
```

Comment:
Please briefly describe what it really does for optimizing RelSubset. For example - passing down the trait requirements of current RelSubset and add converters to the new derived subsets.